library/Freckle/App/Database.hs

```diff
@@ -137,11 +138,41 @@ runDB
   -> m a
 runDB action = do
   pool <- asks getSqlPool
-  Stats.withGauge Stats.dbConnections $
+  gauge <- Stats.lookupGauge Stats.dbConnections
+  let
+    hooks = setAlterBackend defaultSqlPoolHooks $ wrapSqlBackend mempty
+    -- Setting the SqlPoolHooks for metrics collection:
+    -- You may be wondering if this code contains a "double-decrement" bug, because
+    -- perhaps when an exception occurs, both runAfter and runOnException are
+    -- executed. The documentation isn't terribly clear to me on this point, but
+    -- notice:
+    -- https://hackage.haskell.org/package/persistent-2.14.6.0/docs/Database-Persist-SqlBackend-Internal-SqlPoolHooks.html#t:SqlPoolHooks
+    -- especially this part regarding runOnException:
+    -- > This action is performed when an exception is received. The exception is
+    -- > provided as a convenience - it is rethrown once this cleanup function is
+    -- > complete.
+    -- Given that the exception is rethrown, then then typical runAfter handler
+    -- is very likely to not run. Looking at the code to validate:
+    -- https://hackage.haskell.org/package/persistent-2.14.6.0/docs/src/Database.Persist.Sql.Run.html#runSqlPoolWithExtensibleHooks
+    -- we can see that this is indeed how the code operates today -- the
+    -- exception is rethrown, and there aren't any other spots where other code
+    -- might catch the exception that we need to worry about, so in the case of
+    -- an exception, runAfter would not be executed.
+    -- So, this appears to be the intended interpretation.
+    hooks' = hooks
+      { runBefore = \conn mi -> do
+          Stats.incGauge gauge
+          runBefore hooks conn mi
+      , runAfter = \conn mi -> do
+          Stats.decGauge gauge
+          runAfter hooks conn mi
+      , runOnException = \conn mi e -> do
+          Stats.decGauge gauge
+          runOnException hooks conn mi e
+      }
+  Stats.withGauge Stats.dbEnqueuedAndProcessing $
     inSpan "runDB" clientSpanArguments $
-      runSqlPoolWithExtensibleHooks action pool Nothing $
-        setAlterBackend defaultSqlPoolHooks $
-          wrapSqlBackend mempty
+      runSqlPoolWithExtensibleHooks action pool Nothing hooks'
```

Comment:
Oh, kinda sorta. Its related, but tangental. 

I was hoping to not have to do all the mess with the hooks. Via the mechanism I propose above, I _think_ its possible, if I can work out all the details.

Reply:
What I am suggesting is the possiblity of changing all of this:
```haskell
runDB
  :: ( MonadUnliftIO m
     , MonadTracer m
     , MonadReader app m
     , HasSqlPool app
     , HasStatsClient app
     , HasCallStack
     )
  => SqlPersistT m a
  -> m a
runDB action = do
  pool <- asks getSqlPool
  gauge <- Stats.lookupGauge Stats.dbConnections
  let
    hooks = setAlterBackend defaultSqlPoolHooks $ wrapSqlBackend mempty
    -- Setting the SqlPoolHooks for metrics collection:
    -- You may be wondering if this code contains a "double-decrement" bug, because
    -- perhaps when an exception occurs, both runAfter and runOnException are
    -- executed. The documentation isn't terribly clear to me on this point, but
    -- notice:
    -- https://hackage.haskell.org/package/persistent-2.14.6.0/docs/Database-Persist-SqlBackend-Internal-SqlPoolHooks.html#t:SqlPoolHooks
    -- especially this part regarding runOnException:
    -- > This action is performed when an exception is received. The exception is
    -- > provided as a convenience - it is rethrown once this cleanup function is
    -- > complete.
    -- Given that the exception is rethrown, then then typical runAfter handler
    -- is very likely to not run. Looking at the code to validate:
    -- https://hackage.haskell.org/package/persistent-2.14.6.0/docs/src/Database.Persist.Sql.Run.html#runSqlPoolWithExtensibleHooks
    -- we can see that this is indeed how the code operates today -- the
    -- exception is rethrown, and there aren't any other spots where other code
    -- might catch the exception that we need to worry about, so in the case of
    -- an exception, runAfter would not be executed.
    -- So, this appears to be the intended interpretation.
    hooks' = hooks
      { runBefore = \conn mi -> do
          Stats.incGauge gauge
          runBefore hooks conn mi
      , runAfter = \conn mi -> do
          Stats.decGauge gauge
          runAfter hooks conn mi
      , runOnException = \conn mi e -> do
          Stats.decGauge gauge
          runOnException hooks conn mi e
      }
  Stats.withGauge Stats.dbEnqueuedAndProcessing $
    inSpan "runDB" clientSpanArguments $
      runSqlPoolWithExtensibleHooks action pool Nothing hooks'
```
to something like this:
```haskell
runDB
  :: ( MonadUnliftIO m
     , MonadTracer m
     , MonadReader app m
     , HasSqlPool app
     , HasStatsClient app
     , HasCallStack
     )
  => SqlPersistT m a
  -> m a
runDB action = do
  pool <- asks getSqlPool
  gauge <- Stats.lookupGauge Stats.dbConnections
  let
    hooks = setAlterBackend defaultSqlPoolHooks $ wrapSqlBackend mempty
    action' =
      bracket (Stats.inc theGauge) (Stats.dec theGauge) action
  Stats.withGauge Stats.dbEnqueuedAndProcessing $
    inSpan "runDB" clientSpanArguments $
      runSqlPoolWithExtensibleHooks action' pool Nothing hooks
```

not sure if that makes sense? so no need to modify `hooks` to `hooks'`, instead wrap `action`.

Reply:
:shrug: I'm too far away from this to know if that will or won't work or why. Godspeed.